dwave/optimization/include/dwave-optimization/nodes/mathematical.hpp

```diff
@@ -54,6 +55,25 @@ struct min {
     constexpr T operator()(const T& x, const T& y) const { return std::min(x, y); }
 };
 
+template <class T>
+struct modulus {
+    constexpr double operator()(const T& x, const T& y) const {
+        // Copy numpy behavior and return 0 for `x % 0`
+        if (y == 0) {
+            return 0;
+        }
+        const double x_double = static_cast<double>(x);
+        const double y_double = static_cast<double>(y);
+        const double remainder = std::fmod(x_double, y_double);
+        // std::fmod result consistent with numpy for same-sign arguments
+        if ((x_double >= 0) == (y_double >= 0))
+            return remainder;
+        // Make result consistent with numpy for different-sign arguments
+        else
+            return remainder + y_double;
+    }
+};
```

Comment:
```python
template <class T>
struct modulus {};

template <>
struct modulus<double> {
    constexpr double operator()(const double& x, const double& y) const {
        // Copy numpy behavior and return 0 for `x % 0`
        if (y == 0) {
            return 0;
        }
        const double remainder = std::fmod(x, y);

        if ((x >= 0) == (y >= 0)) {
            // std::fmod result consistent with numpy for same-sign arguments
            return remainder;
        } else {
             // Make result consistent with numpy for different-sign arguments
            return remainder + y;
        }
    }
};
```
so because we're hardcoding double behaviour, let's do a slightly silly thing and only implement it for `modulus<double>` we can specialize for other types later if we need to. It also slightly simplifies the code.

I also fixed a few brackets for consistency.

Reply:
Sounds good. It also looks like I need to implement a few of the `BinaryOpNode` functions to fix the gcc tests.

Reply:
Yes, specifically `::min()` , `::max()`, and `::integral()`. Luckily they aren't too complex in this case. Watch the "negative" modulus possibility though.